### PR TITLE
FEATURE: Support `[description|attachment](upload://<short-sha>)` in MD.

### PR DESCRIPTION
**Pull request by tgxworld**

### CODE REVIEW

**Comment by tgxworld** [comment]
Will need to fix the tests tomorrow. We're applying callbacks to pretender incorrectly.

File: app/assets/javascripts/pretty-text/engines/discourse-markdown-it.js.es6 [hunk 1]
```diff
@@ -187,6 +188,26 @@ function setupImageDimensions(md) {
   md.renderer.rules.image = renderImage;
 }
 
+function renderAttachment(tokens, idx, options, env, slf) {
+  const linkOpenToken = tokens[idx];
+  const linkTextToken = tokens[idx + 1];
+  const split = linkTextToken.content.split("|");
```
Comment (jjaffeux) [inline app/assets/javascripts/pretty-text/engines/discourse-markdown-it.js.es6]: Got to be careful on this:

```
"".split("|") // [""]
```

A safer way is:
```
"".split("|").filter(x => x) // []
```

**Comment by eviltrout** [comment]
LGTM, nice :+1:

**Comment by pmusaraj** [comment]
Hmm, let's make sure this works with S3 uploads before merging, it doesn't work on my local S3 set up.

**Comment by tgxworld** [comment]
do you mean the short-url route doesn’t work with S3?

**Comment by pmusaraj** [comment]
Yes, the short-url route gives me a 404 for an upload on S3. The controller isn't finding the upload. But also, `send_file` won't work with a remote URL, I think we'll need to do a redirect here for S3 uploads. (I ran into something similar while working on https://github.com/discourse/discourse/pull/7551/files#diff-60328924b5c773b42aacabc0b5698af4L89).

**Comment by tgxworld** [comment]
O good catch :) let me update this PR and the existing tests to make sure S3 uploads work.

File: app/assets/javascripts/pretty-text/engines/discourse-markdown-it.js.es6 (hunk 1 above)
Comment (tgxworld) [reply app/assets/javascripts/pretty-text/engines/discourse-markdown-it.js.es6]: Thanks for the tip. :+1: I'm checking for the length of `split` in the following code so I don't need an empty array.

File: spec/components/pretty_text_spec.rb [hunk 2]
```diff
@@ -1290,17 +1296,25 @@ def strip_image_wrapping(html)
         </li>
         </ul>
         <p><img src="#{upload.url}" alt="upload"></p>
```
Comment (SamSaffron) [inline spec/components/pretty_text_spec.rb]: I think it is worth explicitly configuring a CDN here so it is clear that the links that come back are not CDN-ified.

Comment (tgxworld) [reply spec/components/pretty_text_spec.rb]: It is impossible for `Upload#url` to return a CDN-ified URL here.

Comment (SamSaffron) [reply spec/components/pretty_text_spec.rb]: Sorry I mean set the global setting so the site has a CDN this confirms that when cooking you never cdnify things.

Comment (tgxworld) [reply spec/components/pretty_text_spec.rb]: OK done :+1:

**Comment by SamSaffron** [comment]
@pmusaraj if this is looking good to you, please merge it in when you start your day, I had a look at the tests here and its a :+1:  from me! You need to build on this so it makes sense that you merge it.

**Comment by tgxworld** [comment]
test failure is not related

**Comment by pmusaraj** [comment]
This looks awesome! 

One tiny thing: do we need to amend the nginx config to skip caching headers for `uploads/short-url` requests?

**Pull request merged**